.github/workflows/ci.yaml

```diff
@@ -37,7 +43,186 @@ jobs:
         with:
           toolchain: stable
           override: true
-      - uses: actions-rs/cargo@v1
+      - name: cargo fetch
+        uses: actions-rs/cargo@v1
+        with:
+          command: fetch
+      - name: cargo test build
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --tests --release
+      - name: cargo test
+        uses: actions-rs/cargo@v1
         with:
           command: test
-          args: --all-features
+          args: --release
+
+  deny-check:
+    name: cargo-deny check
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - name: download cargo-deny
+      shell: bash
+      env:
+        DVS: 0.3.0-beta
+        DREPO: EmbarkStudios/cargo-deny
+        TARGET: x86_64-unknown-linux-musl
+      run: |
+        temp_archive=$(mktemp --suffix=.tar.gz)
+        curl -L --output "$temp_archive" https://github.com/$DREPO/releases/download/$DVS/cargo-deny-$DVS-$TARGET.tar.gz
+
+        tar -xzvf "$temp_archive" -C . --strip-components=1 --wildcards "*/cargo-deny"
+    - name: cargo-deny check licenses
+      run: ./cargo-deny -L debug check license
+    - name: cargo-deny check bans
+      run: ./cargo-deny -L debug check ban
+
+  publish-check:
+    name: Publish Check
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - uses: actions-rs/toolchain@v1
+      with:
+        toolchain: stable
+        override: true
+    - name: cargo fetch
+      uses: actions-rs/cargo@v1
+      with:
+        command: fetch
+    - name: copy README
+      shell: bash
+      run: |
+        cp README.md lib
+        cp README.md cli
+    - name: cargo publish lib
+      uses: actions-rs/cargo@v1
+      with:
+        command: publish
+        args: --dry-run --allow-dirty --manifest-path lib/Cargo.toml
+    - name: cargo publish cli
+      uses: actions-rs/cargo@v1
+      with:
+        command: publish
+        args: --dry-run --allow-dirty --manifest-path cli/Cargo.toml
+
+  publish:
+    name: Publish
+    needs: [test, deny-check, publish-check]
+    runs-on: ubuntu-latest
+    if: startsWith(github.ref, 'refs/tags/')
+    steps:
+    - uses: actions/checkout@v1
+    - uses: actions-rs/toolchain@v1
+      with:
+        toolchain: stable
+        override: true
+    - name: cargo fetch
+      uses: actions-rs/cargo@v1
+      with:
+        command: fetch
+    - name: copy README
+      shell: bash
+      run: |
+        cp README.md lib
+        cp README.md cli
+    - name: cargo publish lib
+      uses: actions-rs/cargo@v1
+      env:
+        CARGO_REGISTRY_TOKEN: ${{ secrets.CRATES_IO_TOKEN }}
+      with:
+        command: publish
+        args: --allow-dirty --manifest-path lib/Cargo.toml
+    - name: cargo publish cli
+      uses: actions-rs/cargo@v1
+      env:
+        CARGO_REGISTRY_TOKEN: ${{ secrets.CRATES_IO_TOKEN }}
+      with:
+        command: publish
+        args: --allow-dirty --manifest-path cli/Cargo.toml
```

Comment:
Is `-allow-dirty` needed? what modifications are done on the repo?

Reply:
Yes, both the lib and cli crates use README.md when publishing to crates.io, but cargo doesn't allow referencing the readme in parent directories, so we copy the README from the root into the crate root when publishing.

Reply:
Oh interesting, and should be fine to use it on CI at least, not having `--allow-dirty` is mostly important for local use so one doesn't have any local changes